akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala

```diff
@@ -191,52 +192,78 @@ object Framing {
+      private var buffer = ByteString.empty
+      private var frameSize = Int.MaxValue
 
-    override def onUpstreamFinish(ctx: Context[ByteString]): TerminationDirective =
-      if (buffer.nonEmpty) ctx.absorbTermination()
-      else ctx.finish()
-
-    private def doParse(ctx: Context[ByteString]): SyncDirective = {
-      def emitFrame(ctx: Context[ByteString]): SyncDirective = {
-        val parsedFrame = buffer.take(frameSize).compact
+      /**
+       * push, and reset frameSize and buffer
+       *
+       */
+      private def pushFrame() = {
+        val emit = buffer.take(frameSize).compact
         buffer = buffer.drop(frameSize)
         frameSize = Int.MaxValue
-        if (ctx.isFinishing && buffer.isEmpty) ctx.pushAndFinish(parsedFrame)
-        else ctx.push(parsedFrame)
+        push(out, emit)
+        if (buffer.size == 0 && isClosed(in)) {
+          completeStage()
+        }
       }
 
-      val bufSize = buffer.size
-      if (bufSize >= frameSize) emitFrame(ctx)
-      else if (bufSize >= minimumChunkSize) {
-        val parsedLength = intDecoder(buffer.iterator.drop(lengthFieldOffset), lengthFieldLength)
-        frameSize = parsedLength + minimumChunkSize
-        if (frameSize > maximumFrameLength)
-          ctx.fail(new FramingException(s"Maximum allowed frame size is $maximumFrameLength but decoded frame header reported size $frameSize"))
-        else if (bufSize >= frameSize) emitFrame(ctx)
-        else tryPull(ctx)
-      } else tryPull(ctx)
-    }
+      /**
+       * try to push downstream, if failed then try to pull upstream
+       *
+       */
+      private def tryPushFrame() = {
+        val buffSize = buffer.size
+        if (buffSize >= frameSize) {
+          pushFrame()
+        } else if (buffSize >= minimumChunkSize) {
+          val parsedLength = intDecoder(buffer.iterator.drop(lengthFieldOffset), lengthFieldLength)
+          frameSize = parsedLength + minimumChunkSize
+          if (frameSize > maximumFrameLength) {
+            failStage(new FramingException(s"Maximum allowed frame size is $maximumFrameLength but decoded frame header reported size $frameSize"))
+          } else if (buffSize >= frameSize) {
+            pushFrame()
+          } else tryPull
+        } else tryPull
+      }
 
-    override def postStop(): Unit = buffer = null
+      private def tryPull = {
+        if (isClosed(in)) {
+          failStage(new FramingException("Stream finished but there was a truncated final frame in the buffer"))
+        } else pull(in)
+      }
+
+      override def onPush(): Unit = {
+        buffer ++= grab(in)
+        tryPushFrame()
+      }
+
+      override def onPull() = tryPushFrame()
+
+      override def onUpstreamFinish(): Unit = {
+        val buffSize = buffer.size
+        if (buffSize == 0) {
+          completeStage()
+        } else if (isAvailable(out) && buffSize > 0) {
```

Comment:
`buffSize` cannot be negative because of the if on line `258`, so we already know the buffer is `nonEmpty` here. Check with `buffer.isEmpty` in the first if and skip the value.